### DISCUSSION TITLE
Move TestResultService to an appropriate kind of background job

### DISCUSSION THREAD

**Discussion by mejtika** [DevAdventCalendar/DevAdventCalendar#396] [Q&A] [answered]
Good Morning @DevAdventCalendar/dac-mentoring-group @DevAdventCalendar/developers,

I need your help in choosing the right solution for a task. A task is to refactor existing TestResultService, which responsibility boils down to calculate points for users. It is now fired manually after each week of Advent. The idea is to move it into some kind of background job and with that comes my question. 
**What kind of background job do you think would be the most appropriate solution for this?**
I'm considering a few options.
1. Hosted service or Worker service
2. Hangfire, Quartz.NET
3. Azure functions

Ok, so after digging deeper into this topic I have some thoughts that I like to share with you.
**Hosted services** provide a clean pattern for executing code within an ASP.NET Core application that sits outside of the request pipeline. Each service is started as an asynchronous background task. Hosted services are useful when e.g. you need to run some background activity within a web server process, you need to pull data from an external service to update a cash or data store, or when responding to messages or events from an external queue or service bus.
**Background services** on the other hand, are simply console applications that can be used to run any kind of background workload where there is no user interface or direct user interaction. Worker services are suitable for long-running processes, for things that are continually being running in the background waiting for something to happen and when it does it immediately jumps in and starts doing some works.

In our case, as we want to run TestResultService only a few times in December each year, the use of hosted services will be inefficient. For hosted service to work in that case, we would have to have in it an infinite while loop and delay for a period of time after each run. Then at the start of the loop, we would have to check the time (and any other conditions) to decide to run the processing workload, or simply skip to the next delay.

So while it's possible to create a "timed" background service, existing libraries, that can perform background workload with a handy interface to schedule it, provides a far more robust solution.
[Coravel](https://docs.coravel.net/), which claims itself as a near-zero config .net core library that makes task scheduling and many more a breeze could be one solution. It leverages Invocables, which represent a self-contained job within the application, to make code much easier to write, compose and maintain.
[Hangfire](https://www.hangfire.io/), which provides an easy way to perform background processing that can be scheduled. Unlike Coravel it is backed by persistent storage and offers a dashboard, where the scheduled jobs can be seen.
I think that with these two we can easily schedule our TestResultService work, using CRON expressions, to be performed in advent time.
Coravel is more lightweight, as it doesn't support persistent storage and dashboard, so Hangfire is more reliable. Once the job is created, it takes responsibility to perform it. It will also log the exceptions and gonna retry each of those jobs adding an increasing delay between them. It is also possible to see, cancel and trigger a job from the dashboard level.

We could also think about handing background processes with Azure functions. They are designed to use with Azure, so I think that we'll need to connect/deploy there. They are also not free, we'll be charged for using them. I think that they can be overkill for our needs.

I would consider doing that task with the use of Coravel/Hangfire. I think there are the most appropriate for our needs and can be also used for other issues like #372. What are your thoughts, DAC people? :D

**Comment by WTobor** [comment]
Good evening ;) can't wait for your thoughts! :)

**Comment by mejtika** [comment]
I've posted my thoughts @DevAdventCalendar/dac-mentoring-group, @DevAdventCalendar/developers!

**Comment by WTobor** [comment, marked as answer]
@patrikulus what do you think about Azure Function? I think that should be the easiest solution. But maybe not the best? :)

**Reply by patrikulus** [reply]
I think this is all overengineering. We already have a pipeline which performs the points counting. All we need is setup a cron trigger for this job. And that's all ;)

**Reply by mejtika** [reply]
@patrikulus So the same story will be in the case of this issue #372, where we want to add some kind of job that will be checking periodically user inactivity and then taking some action? We don't need any of these methods described above?

**Reply by WTobor** [reply]
@patrikulus maybe we should consider using yml ;p

We need o change BinaryPath - we are using .NET Core 3.1, not 2.2
